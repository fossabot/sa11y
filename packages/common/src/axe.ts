/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';

export const axeRuntimeExceptionMsgPrefix = 'Error running accessibility checks using axe:';

// TODO (refactor): Find a way to declare version into axe namespace
// axe.version doesn't exist in typedef https://github.com/dequelabs/axe-core/issues/2378
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
// eslint-disable-next-line import/namespace
export const axeVersion: string | undefined = axe.version;

/**
 * Interface that represents a function that runs axe and returns violations
 */
interface AxeRunner {
    (): Promise<axe.Result[]>;
}

/**
 * Get violations by running axe with given function
 * @param axeRunner - function satisfying AxeRunner interface
 */
export async function getViolations(axeRunner: AxeRunner): Promise<axe.Result[]> {
    let violations;
    try {
        violations = await axeRunner();
    } catch (e) {
        throw new Error(`${axeRuntimeExceptionMsgPrefix} ${e}`);
    }
    return violations;
}
