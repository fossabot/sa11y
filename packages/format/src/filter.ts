/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { Result } from 'axe-core';

/**
 * Filter to post-process a11y results from axe
 */
export interface Filter {
    (violations: Result[], ...args: never[]): Result[];
}

// TODO (refactor): constrain rule id to known rule ids e.g using string literal, keyof, in etc
//  e.g. https://stackoverflow.com/a/54061487
// const ruleIDs = getRules().map((ruleObj) => ruleObj.ruleId);
// type ruleID = keyof ruleIDs;
// type ruleID = typeof ruleIDs[number];
type ruleID = string;
type cssSelectors = string[];

/**
 * Exception list of map of rule to corresponding css targets that needs to be filtered from a11y results.
 */
type exceptionList = Record<ruleID, cssSelectors>;

/**
 * Filter a11y violations from axe based on given {@link exceptionList}
 * @param violations - List of violations found with axe
 * @param exceptionList - {@link exceptionList} of map of rule to corresponding css targets that needs to be filtered from a11y results
 */
export function exceptionListFilter(violations: Result[], exceptionList: exceptionList = {}): Result[] {
    // TODO (refactor): Use map, filter instead of for loops? without compromising readability/debugging-ability
    //  is map/filter more performant? without having to create an intermediary results ourselves?
    // return violations.filter((violation) =>
    //     Object.keys(exceptionList).includes(violation.id)
    //         ? violation.nodes.filter(
    //               (result) => !result.target.filter((cssSelector) => !exceptionList[violation.id].includes(cssSelector))
    //           ).length > 0
    //         : true
    // );

    // TODO (refactor): Is there a better way to deal with this condition within the flow of the loop below?
    if (!(Object.entries(exceptionList).length > 0)) return violations;

    const filteredViolations: Result[] = [];

    for (const violation of violations) {
        for (const [ruleID, cssSelectors] of Object.entries(exceptionList)) {
            if (violation.id !== ruleID) {
                filteredViolations.push(violation);
            } else {
                for (const result of violation.nodes) {
                    const filteredResults = result.target.filter((cssSelector) => !cssSelectors.includes(cssSelector));
                    if (filteredResults.length > 0) {
                        filteredViolations.push(violation);
                    }
                }
            }
        }
    }

    return filteredViolations;
}
