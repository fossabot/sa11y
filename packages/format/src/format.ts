/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Result } from 'axe-core';
import * as Handlebars from 'handlebars';
import fs from 'fs';
import path from 'path';

/**
 * Formatter defines the function signature to format accessibility violations found by axe
 */
export interface Formatter {
    (violations: Result[]): string;
}

/**
 * Read template from filesystem and pre-compile it
 */
export async function readTemplate() {
    // TODO (type):  error TS2304: Cannot find name 'TemplateSpecification'
    // export async function readTemplate(): Promise<TemplateSpecification> {
    return new Promise((resolve, reject) => {
        // TODO (Debug): Loading in compiled template file instead without needing to use 'fs' module doesn't work
        // TODO (script): Add run script to generate compiled template: "yarn handlebars `pwd`/template.mustache -f `pwd`/template.compiled.js --knownOnly"
        // return require('./template.compiled.js');
        // TODO (Refactor): When bundling for browser use https://handlebarsjs.com/installation/integrations.html
        fs.readFile(path.resolve(__dirname, './template.mustache'), (err, data) => {
            if (err) reject(err);
            resolve(Handlebars.precompile(data.toString()));
        });
    });
}

/**
 * Format accessibility results from axe
 * @param violations - Result list from axe
 * */
// TODO: Add handlebars template for formatting
// TODO: Add support for different output formats console(colored), plain text, HTML, xUnit
export function a11yResultsFormatter(violations: Result[]): string {
    return violations
        .map((violation) => {
            return violation.nodes
                .map((node) => {
                    const selectors = node.target.join(', ');
                    return `${violation.help} (${violation.id}): ${selectors}\n  - More info: ${violation.helpUrl}`;
                })
                .join('\n\n');
        })
        .join('\n\n');
}
