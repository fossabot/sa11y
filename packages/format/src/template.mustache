{{!-- Mustache template to process list of violations from axe
      Model: com.salesforce.sa11y.AxeResult

      Ref:
        - http://mustache.github.io/mustache.5.html
        - https://github.com/samskivert/jmustache
        - [Special variables](https://github.com/samskivert/jmustache#special-variables)
          used in this template
            * "-index" : special variable containing index of iteration starting from 1
--}}
{{numElements}} HTML elements have accessibility issue(s). {{results.size}} rules failed.
Page URL       :  {{{pageURL}}}
Test mode      :  {{testMode}}

{{#results}}
    ({{-index}}) [{{ruleName}}]
        * Description      : {{{ruleDescription}}}.
        * More info        : {{{helpURL}}}
        {{#wcagCriteria}}
        * WCAG criteria    : {{wcagCriteria}}
        {{/wcagCriteria}}
        * Error element(s) : {{elements.size}}
        {{#elements}}
            ({{-index}}) HTML element : {{{html}}}
              - CSS selector(s)    :  {{{target}}}
            {{#additionalInfo}}
              - Additional info    : {{{additionalInfo}}}
            {{/additionalInfo}}
            {{#relatedElements}}
              - Related element(s) :
                {{#relatedElements}}
                * HTML             : {{{html}}}
                  CSS selector(s)  : {{{target}}}
                {{/relatedElements}}
            {{/relatedElements}}
        {{/elements}}
{{/results}}
