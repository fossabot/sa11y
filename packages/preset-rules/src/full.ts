/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { getA11yConfig } from './a11yConfig';
import * as axe from 'axe-core';

export const full = getA11yConfig(axe.getRules().map((ruleObj) => ruleObj.ruleId));
