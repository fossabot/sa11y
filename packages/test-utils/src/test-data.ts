/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import fs from 'fs';
import path from 'path';

// DOM with a11y issues
export const domWithA11yIssuesBodyID = 'dom-with-issues';
export const htmlFileWithA11yIssues = path.resolve(__dirname, '../__data__/a11yIssues.html');
export const domWithA11yIssues = fs.readFileSync(htmlFileWithA11yIssues).toString();

// DOM containing no a11y issues
export const shadowDomID = 'upside-down';
export const htmlFileWithNoA11yIssues = path.resolve(__dirname, '../__data__/noA11yIssues.html');
export const domWithNoA11yIssues = fs.readFileSync(htmlFileWithNoA11yIssues).toString();
