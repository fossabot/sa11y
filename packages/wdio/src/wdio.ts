/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
import { A11yConfig, extended } from '@sa11y/preset-rules';
import * as axe from 'axe-core';
import { axeRuntimeExceptionMsgPrefix } from '@sa11y/common';
import { A11yError } from '@sa11y/format';
import { BrowserObject } from 'webdriverio';

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export async function assertAccessible(driver: BrowserObject = browser, rules: A11yConfig = extended) {
    // TODO (deduplicate): with @sa11y/assert
    let violations;
    try {
        // inject the script
        await driver.execute(axe.source);

        // run inside browser and get results
        violations = await driver.executeAsync((rules, done) => {
            // TODO (feat): Add ability to target specific parts of document (by Element or CSS selector)
            axe.run(document, rules, function (err: Error, results: axe.AxeResults) {
                if (err) throw err;
                done(results.violations);
            });
        }, rules);
    } catch (e) {
        throw new Error(`${axeRuntimeExceptionMsgPrefix} ${e}`);
    }
    if (violations.length > 0) {
        throw new A11yError(violations);
    }
}
