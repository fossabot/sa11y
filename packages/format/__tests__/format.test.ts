/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { a11yResultsFormatter } from '..';
import { beforeEachSetup, domWithA11yIssues } from '@sa11y/test-utils';
import { readTemplate } from '../src/format';

beforeEach(beforeEachSetup);

describe('a11y Results Formatter', () => {
    it('should load compiled template', async () => {
        expect(await readTemplate()).toMatchSnapshot();
    });

    it('should format a11y issues as expected', async () => {
        expect.assertions(2);
        document.body.innerHTML = domWithA11yIssues;
        await axe.run(document).then((results) => {
            expect(results).toBeDefined();
            expect(a11yResultsFormatter(results.violations)).toMatchSnapshot();
        });
    });
});
