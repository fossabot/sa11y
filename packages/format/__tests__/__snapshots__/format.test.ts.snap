// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a11y Results Formatter should format a11y issues as expected 1`] = `
"Page must have means to bypass repeated blocks (bypass): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/bypass?application=axeAPI

Documents must have <title> element to aid in navigation (document-title): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/document-title?application=axeAPI

Document must have one main landmark (landmark-one-main): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axeAPI

Links must have discernible text (link-name): a
  - More info: https://dequeuniversity.com/rules/axe/3.5/link-name?application=axeAPI

Page must contain a level-one heading (page-has-heading-one): html
  - More info: https://dequeuniversity.com/rules/axe/3.5/page-has-heading-one?application=axeAPI"
`;

exports[`a11y Results Formatter should load compiled template 1`] = `
"{\\"1\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, options, alias1=depth0 != null ? depth0 : (container.nullContext || {}), alias2=container.hooks.helperMissing, alias3=\\"function\\", alias4=container.escapeExpression, alias5=container.hooks.blockHelperMissing, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    }, buffer = 
  \\"    (\\"
    + alias4(((helper = (helper = lookupProperty(helpers,\\"-index\\") || (depth0 != null ? lookupProperty(depth0,\\"-index\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"-index\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":15,\\"column\\":5},\\"end\\":{\\"line\\":15,\\"column\\":15}}}) : helper)))
    + \\") [\\"
    + alias4(((helper = (helper = lookupProperty(helpers,\\"ruleName\\") || (depth0 != null ? lookupProperty(depth0,\\"ruleName\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"ruleName\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":15,\\"column\\":18},\\"end\\":{\\"line\\":15,\\"column\\":30}}}) : helper)))
    + \\"] \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"helpText\\") || (depth0 != null ? lookupProperty(depth0,\\"helpText\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"helpText\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":15,\\"column\\":32},\\"end\\":{\\"line\\":15,\\"column\\":46}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n    * Description      : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"ruleDescription\\") || (depth0 != null ? lookupProperty(depth0,\\"ruleDescription\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"ruleDescription\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":16,\\"column\\":25},\\"end\\":{\\"line\\":16,\\"column\\":46}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n    * Help URL         : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"helpURL\\") || (depth0 != null ? lookupProperty(depth0,\\"helpURL\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"helpURL\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":17,\\"column\\":25},\\"end\\":{\\"line\\":17,\\"column\\":38}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n    * More info        : \\"
    + alias4(((helper = (helper = lookupProperty(helpers,\\"presetRules\\") || (depth0 != null ? lookupProperty(depth0,\\"presetRules\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"presetRules\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":18,\\"column\\":25},\\"end\\":{\\"line\\":18,\\"column\\":40}}}) : helper)))
    + \\" ruleset\\";
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"wcagCriteria\\") || (depth0 != null ? lookupProperty(depth0,\\"wcagCriteria\\") : depth0)) != null ? helper : alias2),(options={\\"name\\":\\"wcagCriteria\\",\\"hash\\":{},\\"fn\\":container.program(2, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":18,\\"column\\":48},\\"end\\":{\\"line\\":18,\\"column\\":100}}}),(typeof helper === alias3 ? helper.call(alias1,options) : helper));
  if (!lookupProperty(helpers,\\"wcagCriteria\\")) { stack1 = alias5.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  buffer += \\"\\\\n    * Error element(s) : \\"
    + alias4(container.lambda(((stack1 = (depth0 != null ? lookupProperty(depth0,\\"elements\\") : depth0)) != null ? lookupProperty(stack1,\\"size\\") : stack1), depth0))
    + \\"\\\\n\\";
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"elements\\") || (depth0 != null ? lookupProperty(depth0,\\"elements\\") : depth0)) != null ? helper : alias2),(options={\\"name\\":\\"elements\\",\\"hash\\":{},\\"fn\\":container.program(4, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":20,\\"column\\":4},\\"end\\":{\\"line\\":33,\\"column\\":17}}}),(typeof helper === alias3 ? helper.call(alias1,options) : helper));
  if (!lookupProperty(helpers,\\"elements\\")) { stack1 = alias5.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  return buffer;
},\\"2\\":function(container,depth0,helpers,partials,data) {
    var helper, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    };

  return \\", \\"
    + container.escapeExpression(((helper = (helper = lookupProperty(helpers,\\"wcagCriteria\\") || (depth0 != null ? lookupProperty(depth0,\\"wcagCriteria\\") : depth0)) != null ? helper : container.hooks.helperMissing),(typeof helper === \\"function\\" ? helper.call(depth0 != null ? depth0 : (container.nullContext || {}),{\\"name\\":\\"wcagCriteria\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":18,\\"column\\":67},\\"end\\":{\\"line\\":18,\\"column\\":83}}}) : helper)));
},\\"4\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, options, alias1=depth0 != null ? depth0 : (container.nullContext || {}), alias2=container.hooks.helperMissing, alias3=\\"function\\", alias4=container.hooks.blockHelperMissing, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    }, buffer = 
  \\"        (\\"
    + container.escapeExpression(((helper = (helper = lookupProperty(helpers,\\"-index\\") || (depth0 != null ? lookupProperty(depth0,\\"-index\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"-index\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":21,\\"column\\":9},\\"end\\":{\\"line\\":21,\\"column\\":19}}}) : helper)))
    + \\") HTML element : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"html\\") || (depth0 != null ? lookupProperty(depth0,\\"html\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"html\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":21,\\"column\\":36},\\"end\\":{\\"line\\":21,\\"column\\":46}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n        - CSS selector(s)    :  \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"target\\") || (depth0 != null ? lookupProperty(depth0,\\"target\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"target\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":22,\\"column\\":32},\\"end\\":{\\"line\\":22,\\"column\\":44}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n\\";
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"additionalInfo\\") || (depth0 != null ? lookupProperty(depth0,\\"additionalInfo\\") : depth0)) != null ? helper : alias2),(options={\\"name\\":\\"additionalInfo\\",\\"hash\\":{},\\"fn\\":container.program(5, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":23,\\"column\\":8},\\"end\\":{\\"line\\":25,\\"column\\":27}}}),(typeof helper === alias3 ? helper.call(alias1,options) : helper));
  if (!lookupProperty(helpers,\\"additionalInfo\\")) { stack1 = alias4.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"relatedElements\\") || (depth0 != null ? lookupProperty(depth0,\\"relatedElements\\") : depth0)) != null ? helper : alias2),(options={\\"name\\":\\"relatedElements\\",\\"hash\\":{},\\"fn\\":container.program(7, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":26,\\"column\\":8},\\"end\\":{\\"line\\":32,\\"column\\":28}}}),(typeof helper === alias3 ? helper.call(alias1,options) : helper));
  if (!lookupProperty(helpers,\\"relatedElements\\")) { stack1 = alias4.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  return buffer;
},\\"5\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    };

  return \\"            - Additional info    : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"additionalInfo\\") || (depth0 != null ? lookupProperty(depth0,\\"additionalInfo\\") : depth0)) != null ? helper : container.hooks.helperMissing),(typeof helper === \\"function\\" ? helper.call(depth0 != null ? depth0 : (container.nullContext || {}),{\\"name\\":\\"additionalInfo\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":24,\\"column\\":35},\\"end\\":{\\"line\\":24,\\"column\\":55}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n\\";
},\\"7\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, options, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    }, buffer = 
  \\"            - Related element(s) :\\\\n\\";
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"relatedElements\\") || (depth0 != null ? lookupProperty(depth0,\\"relatedElements\\") : depth0)) != null ? helper : container.hooks.helperMissing),(options={\\"name\\":\\"relatedElements\\",\\"hash\\":{},\\"fn\\":container.program(8, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":28,\\"column\\":12},\\"end\\":{\\"line\\":31,\\"column\\":32}}}),(typeof helper === \\"function\\" ? helper.call(depth0 != null ? depth0 : (container.nullContext || {}),options) : helper));
  if (!lookupProperty(helpers,\\"relatedElements\\")) { stack1 = container.hooks.blockHelperMissing.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  return buffer;
},\\"8\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, alias1=depth0 != null ? depth0 : (container.nullContext || {}), alias2=container.hooks.helperMissing, alias3=\\"function\\", lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    };

  return \\"                * HTML             : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"html\\") || (depth0 != null ? lookupProperty(depth0,\\"html\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"html\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":29,\\"column\\":37},\\"end\\":{\\"line\\":29,\\"column\\":47}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n                CSS selector(s)  : \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"target\\") || (depth0 != null ? lookupProperty(depth0,\\"target\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"target\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":30,\\"column\\":35},\\"end\\":{\\"line\\":30,\\"column\\":47}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n\\";
},\\"compiler\\":[8,\\">= 4.3.0\\"],\\"main\\":function(container,depth0,helpers,partials,data) {
    var stack1, helper, options, alias1=depth0 != null ? depth0 : (container.nullContext || {}), alias2=container.hooks.helperMissing, alias3=\\"function\\", alias4=container.escapeExpression, lookupProperty = container.lookupProperty || function(parent, propertyName) {
        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
          return parent[propertyName];
        }
        return undefined
    }, buffer = 
  alias4(((helper = (helper = lookupProperty(helpers,\\"numElements\\") || (depth0 != null ? lookupProperty(depth0,\\"numElements\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"numElements\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":11,\\"column\\":0},\\"end\\":{\\"line\\":11,\\"column\\":15}}}) : helper)))
    + \\" HTML elements have accessibility issue(s). \\"
    + alias4(container.lambda(((stack1 = (depth0 != null ? lookupProperty(depth0,\\"results\\") : depth0)) != null ? lookupProperty(stack1,\\"size\\") : stack1), depth0))
    + \\" rules failed.\\\\nPage URL       :  \\"
    + ((stack1 = ((helper = (helper = lookupProperty(helpers,\\"pageURL\\") || (depth0 != null ? lookupProperty(depth0,\\"pageURL\\") : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{\\"name\\":\\"pageURL\\",\\"hash\\":{},\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":12,\\"column\\":18},\\"end\\":{\\"line\\":12,\\"column\\":31}}}) : helper))) != null ? stack1 : \\"\\")
    + \\"\\\\n\\\\n\\";
  stack1 = ((helper = (helper = lookupProperty(helpers,\\"results\\") || (depth0 != null ? lookupProperty(depth0,\\"results\\") : depth0)) != null ? helper : alias2),(options={\\"name\\":\\"results\\",\\"hash\\":{},\\"fn\\":container.program(1, data, 0),\\"inverse\\":container.noop,\\"data\\":data,\\"loc\\":{\\"start\\":{\\"line\\":14,\\"column\\":0},\\"end\\":{\\"line\\":34,\\"column\\":12}}}),(typeof helper === alias3 ? helper.call(alias1,options) : helper));
  if (!lookupProperty(helpers,\\"results\\")) { stack1 = container.hooks.blockHelperMissing.call(depth0,stack1,options)}
  if (stack1 != null) { buffer += stack1; }
  return buffer + \\"\\\\nFor more info about automated accessibility testing: https://sfdc.co/a11y-test\\\\nFor tips on fixing accessibility bugs: https://sfdc.co/a11y\\\\nQuestions? Post on Accessibility Team Chatter: https://sfdc.co/a11y-gus\\\\n\\";
},\\"useData\\":true}"
`;
